### PR TITLE
Fixed TF_DOUBLE not granting hit bonus when triggered

### PR DESCRIPTION
**Pull request by csnv**
### Pull Request Prelude

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
Added hit bonus to TF_DOUBLE as explained [here](https://irowiki.org/classic/Double_Attack)

**Issues addressed:** 

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/battle.c [hunk 1]
```diff
@@ -4855,15 +4860,15 @@ static struct Damage battle_calc_weapon_attack(struct block_list *src, struct bl
...
 		 || ( sc && sc->data[SC_KAGEMUSYA] && sd->weapontype1 != W_FIST ) // Need confirmation
 		) {
 			//Success chance is not added, the higher one is used [Skotlex]
-			if( rnd()%100 < ( 5*skill_lv > sd->bonus.double_rate ? 5*skill_lv : sc && sc->data[SC_KAGEMUSYA]?sc->data[SC_KAGEMUSYA]->val1*3:sd->bonus.double_rate ) )
-			{
-				wd.div_ = skill->get_num(TF_DOUBLE,skill_lv?skill_lv:1);
+			if (rnd() % 100 < (5 * skill_lv > sd->bonus.double_rate ? 5 * skill_lv : sc && sc->data[SC_KAGEMUSYA] ? sc->data[SC_KAGEMUSYA]->val1 * 3 : sd->bonus.double_rate)) {
+				wd.div_ = skill->get_num(TF_DOUBLE, skill_lv ? skill_lv : 1);
 				wd.type = BDT_MULTIHIT;
+				hitrate += skill_lv;
```
Comment (skyleo) [inline src/map/battle.c:4871]: That's not how it works in pre-renewal.

The bonus isn't a flat +10 HIT.

It's added up together with the BS_WEAPONRESEARCH bonus and potentially other percentual ones and then the hitrate stemming from own hit, enemy flee, fogwall and other status effects gets this as a percentual bonus.

So it's more like `hitplusperc += skill_lv;` for TF_DOUBLE and for BS_WEAPONRESEARCH it's `hitplusperc += skill_lv * 2;` (with hitplusperc being initially `0`)

And that is applied at the very end of the hitrate calculation. like so:
`hitrate += hitrate * hitplusperc / 100`

(although in aegis it's more like `hitrate = hitrate * (100 + hitplusperc) / 100;`)

**Review by skyleo — CHANGES_REQUESTED** [review]

File: src/map/battle.c [hunk 2]
```diff
@@ -4855,15 +4860,15 @@ static struct Damage battle_calc_weapon_attack(struct block_list *src, struct bl
 
 	if (sd && !skill_id) {
 		//Check for double attack.
-		if (( (skill_lv=pc->checkskill(sd,TF_DOUBLE)) > 0 && sd->weapontype1 == W_DAGGER )
+		if (( (skill_lv = pc->checkskill(sd,TF_DOUBLE)) > 0 && sd->weapontype1 == W_DAGGER )
```
Comment (MrKeiKun) [inline src/map/battle.c:4863]: since you fix this, complete it

Comment (csnv) [reply src/map/battle.c:4863]: Do you mean the spaces around parenthesis? Please elaborate.

Comment (MrKeiKun) [reply src/map/battle.c:4863]: that and `sd, TF_DOUBLE`

File: src/map/battle.c [hunk 3]
```diff
@@ -5106,15 +5108,18 @@ static struct Damage battle_calc_weapon_attack(struct block_list *src, struct bl
 				break;
 		}
 
-		if( sd ) {
+		if (sd) {
```
Comment (skyleo) [inline src/map/battle.c:5111]: In every line that you touched where sd or other struct pointers are used please add != NULL.
One shall not mistreat non-boolean values as boolean values.

**Review by skyleo — CHANGES_REQUESTED** [review]

File: src/map/battle.c [hunk 4]
```diff
@@ -5106,15 +5108,18 @@ static struct Damage battle_calc_weapon_attack(struct block_list *src, struct bl
...
-			if ((temp = pc->checkskill(sd,BS_WEAPONRESEARCH)) > 0)
-				hitrate += hitrate * ( 2 * temp ) / 100;
+			if ((temp = pc->checkskill(sd, BS_WEAPONRESEARCH)) > 0)
+				hitpercbonus += 2 * temp;
 
 			if ((sd->weapontype == W_1HSWORD || sd->weapontype == W_DAGGER) && (temp = pc->checkskill(sd, GN_TRAINING_SWORD)) > 0)
 				hitrate += 3 * temp;
 		}
 
+		if (hitpercbonus > 0)
```
Comment (skyleo) [inline src/map/battle.c:5120]: I think it's fine if it's negative too. May have seen negative hitpercbonus somewhere in Aegis.

**Review by skyleo — CHANGES_REQUESTED** [review]

File: src/map/battle.c (hunk 4 above)
Comment (csnv) [reply src/map/battle.c:5120]: Currently there's no way for hitpercbonus to be negative and including negative values might be confusing for whoever reads it in the future. I'd rather keep it like that until a case that forces it to be negative is included.

File: src/map/battle.c [hunk 5]
```diff
@@ -4855,15 +4856,16 @@ static struct Damage battle_calc_weapon_attack(struct block_list *src, struct bl
...
-		 || ( sd->bonus.double_rate > 0 && sd->weapontype1 != W_FIST ) //Will fail bare-handed
-		 || ( sc && sc->data[SC_KAGEMUSYA] && sd->weapontype1 != W_FIST ) // Need confirmation
+		if (((skill_lv = pc->checkskill(sd, TF_DOUBLE)) > 0 && sd->weapontype1 == W_DAGGER)
+		 || (sd->bonus.double_rate > 0 && sd->weapontype1 != W_FIST) //Will fail bare-handed
+		 || (sc && sc->data[SC_KAGEMUSYA] && sd->weapontype1 != W_FIST) // Need confirmation
 		) {
-			//Success chance is not added, the higher one is used [Skotlex]
-			if( rnd()%100 < ( 5*skill_lv > sd->bonus.double_rate ? 5*skill_lv : sc && sc->data[SC_KAGEMUSYA]?sc->data[SC_KAGEMUSYA]->val1*3:sd->bonus.double_rate ) )
+			// Success chance is not added, the higher one is used [Skotlex]
+			if (rnd() % 100 < (5 * skill_lv > sd->bonus.double_rate ? 5 * skill_lv : sc && sc->data[SC_KAGEMUSYA] ? sc->data[SC_KAGEMUSYA]->val1 * 3 : sd->bonus.double_rate))
```
Comment (MrKeiKun) [inline src/map/battle.c:4864]: `sc != NULL && sc->data[SC_KAGEMUSYA] != NULL`

File: src/map/battle.c [hunk 6]
```diff
@@ -5068,20 +5070,20 @@ static struct Damage battle_calc_weapon_attack(struct block_list *src, struct bl
...
 				break;
 			case KN_PIERCE:
 			case ML_PIERCE:
-				hitrate += hitrate * 5 * skill_lv / 100;
+				hitpercbonus += 5 * skill_lv;
 				break;
 			case AS_SONICBLOW:
-				if(sd && pc->checkskill(sd,AS_SONICACCEL)>0)
-					hitrate += hitrate * 50 / 100;
+				if (sd && pc->checkskill(sd, AS_SONICACCEL) > 0)
```
Comment (MrKeiKun) [inline src/map/battle.c:5080]: `sd != NULL`

File: src/map/battle.c [hunk 7]
```diff
@@ -5068,20 +5070,20 @@ static struct Damage battle_calc_weapon_attack(struct block_list *src, struct bl
...
+					hitpercbonus += 50;
 				break;
 			case MC_CARTREVOLUTION:
 			case GN_CART_TORNADO:
 			case GN_CARTCANNON:
-				if( sd && pc->checkskill(sd, GN_REMODELING_CART) )
+				if (sd && pc->checkskill(sd, GN_REMODELING_CART))
```
Comment (MrKeiKun) [inline src/map/battle.c:5086]: `sd != NULL`

**Review by MrKeiKun — CHANGES_REQUESTED** [review]

**Review by skyleo — APPROVED** [review]
LGTM now. 

good job.

**Review by MishimaHaruna — APPROVED** [review]

**Pull request merged**